### PR TITLE
Fixed #26142 -- Added a 'edit only' mode for model formsets

### PR DESCRIPTION
**Pull request by Parth1811**
ticket-26142

This patch introduces an 'edit only' mode in the BaseModelFormSet
which blocks the creation of new models and provides a more robust way
to do so.

### CODE REVIEW

**Comment by felixxm** [comment]
@Parth1811 Thanks for this patch :+1: Docs and release notes are missing, moreover please take a look at [comments](https://github.com/django/django/pull/6053#issuecomment-175872039) from the previous attempt.

**Comment by Parth1811** [comment]
Hey @felixxm,

From the issues mentioned in the above linked comment, my patch successfully addresses the first two points.

> It needs to support deletion of the given forms
> It needs to ensure that created forms cannot take the place of deleted forms

Since I'm blocking the creation of new forms, it won't affect the deletion of the forms in any way. Moreover since its a hard condition even if one deletes a form no new forms can be created.

And about the last point, I'm not sure if there is some edge case where this can happen, but I feel this should also work. But I would just try to find if I'm missing something.

> The ModelForms should check that the same objects are being updated/deleted as were given initially

**Comment by carltongibson** [comment]
Hey @Parth1811.

#### Tests

* For Deletion: I think you're right, it shouldn't be affected but, please can you add a test case showing that this still works. (It serves as a regression test for the future if nothing else.)
* The _cannot create_ test looks ≈OK. 

For _cannot swap_: create two instances, create a formset with the first, but submit using the PK of the second, the second will be edited. 

I can't really see how we can solve this without stashing a (checksum) value in the session or such: in any form submission there are **two** requests, the GET which generates the formset and the POST which processes it. To ensure that the set of PKs in the second request matches that from the first we have to store state... (No? Anyone?) 

That may be valuable but it strikes me as beyond the scope of what's needed here: that formsets can be stopped from **creating** a new instance is an enhancement. We can take it as such (I think) **but** that limitation should be documented. 

#### Docs and Release notes

Please add those as @felixxm points out.

File: tests/model_formsets/tests.py [hunk 1]
```diff
@@ -451,6 +451,41 @@ def test_min_num_with_existing(self):
...
+            'form-0-name': 'Walt Whitman',
+        }
+        formset = AuthorFormSet(data, queryset=Author.objects.all(), edit_only=True)
+        self.assertTrue(formset.is_valid())
+        formset.save()
+        self.assertQuerysetEqual(Author.objects.all(), [
+            '<Author: Charles Baudelaire>',
+        ])
+
+    def test_instance_creation_without_edit_only(self):
```
Comment (carltongibson) [inline tests/model_formsets/tests.py]: Move this test to be first (in the change here). Then we're testing "without edit_onlt", then "with edit_only". 


Remove the comments, I don't think they add too much.

**Comment by Parth1811** [comment]
> For _cannot swap_: create two instances, create a formset with the first, but submit using the PK of the second, the second will be edited.

Hey, @carltongibson  this is not true if the object is not in the queryset of the formset, it won't be edited.

**Comment by carltongibson** [comment]
Hey @Parth1811. "...if the object is not in the queryset of the formset, it won't be edited..." — Ah, well that sounds great. I'll have a closer look but 👍

File: django/forms/models.py [hunk 2]
```diff
@@ -669,7 +670,10 @@ def save_m2m():
                 for form in self.saved_forms:
                     form.save_m2m()
             self.save_m2m = save_m2m
-        return self.save_existing_objects(commit) + self.save_new_objects(commit)
+        if not self.edit_only:
+            return self.save_existing_objects(commit) + self.save_new_objects(commit)
+        else:
+            return self.save_existing_objects(commit)
```
Comment (frnhr) [inline django/forms/models.py]: I think we are duplicating functionality with handling `edit_only` attribute. The save behaviour can already be achieved by adding this to the Inline class:
```  python
    def has_add_permission(self, request, obj):
        return False
```
Assuming I'm not missing something, please let me know.

If overriding this method gives the desired effect, then perhaps we can only check for this new `edit_only` attribute in the default implementation of `has_add_permission` method?

Also, does `edit_only` preclude deleting existing inline instances? If yes, we can do the same check in `has_delete_permission` method.

Comment (frnhr) [reply django/forms/models.py]: Or perhaps follow what has been implemented with `can_delete` attribute. It is as an "and" with `has_delete_permission`: https://github.com/django/django/search?p=1&q=can_delete

Maybe adding a new `can_add` attribute in a similar way would be more consistent with the current code?

**Comment by smithdc1** [comment]
@Parth1811 hope you are well.

Are you still interested in continuing with this ticket?

**Comment by Parth1811** [comment]
@smithdc1 Yes, actually I was awaiting review on this PR and then it went under the radar

File: docs/topics/forms/modelforms.txt [hunk 3]
```diff
@@ -951,11 +951,14 @@ extra forms displayed.
...
     ['Charles Baudelaire', 'Paul Verlaine', 'Walt Whitman']
 
+<<<<<<< HEAD
+=======
 Also, ``extra=0`` doesn't prevent creation of new model instances as you can
 :ref:`add additional forms with JavaScript <understanding-the-managementform>`
 or send additional POST data. Formsets :ticket:`don't yet provide functionality
 <26142>` for an "edit only" view that prevents creation of new instances.
 
+>>>>>>> master
```
Comment (jacobtylerwalls) [inline docs/topics/forms/modelforms.txt:961]: artifacts to remove

File: docs/topics/forms/modelforms.txt [hunk 4]
```diff
@@ -972,6 +975,29 @@ so long as the total number of forms does not exceed ``max_num``::
...
 of forms displayed (1000). In practice this is equivalent to no limit.
 
+Also, ``extra=0`` doesn't prevent creation of new model instances as you can
+:ref:`add additional forms with JavaScript <understanding-the-managementform>`
+or just send additional POST data.
+
+`edit only` mode for the model formsets
+---------------------------------------
+
+.. versionadded:: 3.0
```
Comment (jacobtylerwalls) [inline docs/topics/forms/modelforms.txt:985]: Let's target 4.0

**Review by jacobtylerwalls — COMMENTED** [review]
Hi @Parth1811 thanks for adding docs and for your patience. I'll spruce up the ticket flags to get your patch more visible for a final review. Meanwhile a couple things to clean up when you have a moment. Cheers.

**Comment by carltongibson** [comment]
At a quick look, it seems my points on the tests should be covered by `test_prevent_editing_object_outside_of_queryset` — i.e. we restrict editing on items outside the original queryset. @jacobtylerwalls's tidy-ups need addressing. 

Parth, are you still will us on this one? (If not we could unassign and someone else could pick this up.) 
Thanks.

**Comment by Nakmak98** [comment]
Hello, @carltongibson. I am interested to take this ticket. I tried to run tests and got an error. 
`File "../django/tests/model_formsets/tests.py", line 500, in test_deletion_with_edit_only
    formset = PoetFormSet(data, queryset=Poet.objects.all(), edit_only=True)
  File "../django/venv/lib/python3.9/site-packages/django/forms/models.py", line 573, in __init__
    super().__init__(**{'data': data, 'files': files, 'auto_id': auto_id, 'prefix': prefix, **kwargs})

TypeError: __init__() got an unexpected keyword argument 'edit_only'`

Have these tests passed without errors in the past?

**Comment by carltongibson** [comment]
HI @Nakmak98 — thanks for the interest. 

TBH this is so old I can't recall the details. You'll need to look into the failure. 

The PR should be adding the kwarg so... 🤔

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**